src/miner_sup.erl

```diff
@@ -41,118 +32,14 @@ start_link() ->
-            {ok, ""} -> [];
-            {ok, Seeds} -> string:split(Seeds, ",", all);
-            _ -> []
-        end,
-    SeedNodeDNS = application:get_env(blockchain, seed_node_dns, []),
-    % look up the DNS record and add any resulting addresses to the SeedNodes
-    % no need to do any checks here as any bad combination results in an empty list
-    SeedAddresses = string:tokens(lists:flatten([string:prefix(X, "blockchain-seed-nodes=") || [X] <- inet_res:lookup(SeedNodeDNS, in, txt), string:prefix(X, "blockchain-seed-nodes=") /= nomatch]), ","),
-    Port = application:get_env(blockchain, port, 0),
-    NumConsensusMembers = application:get_env(blockchain, num_consensus_members, 4),
-    BaseDir = application:get_env(blockchain, base_dir, "data"),
-    %% TODO: Remove when cuttlefish
-    MaxInboundConnections = application:get_env(blockchain, max_inbound_connections, 10),
-
-    %% downlink packets from state channels go here
-    application:set_env(blockchain, sc_client_handler, miner_lora),
-
-    case application:get_env(blockchain, key, undefined) of
-        undefined ->
-            #{ pubkey := PublicKey,
-               ecdh_fun := ECDHFun,
-               sig_fun := SigFun
-             } = miner_keys:keys({file, BaseDir}),
-            ECCWorker = [];
-        {ecc, Props} when is_list(Props) ->
-            #{ pubkey := PublicKey,
-               key_slot := KeySlot,
-               ecdh_fun := ECDHFun,
-               sig_fun := SigFun
-             } = miner_keys:keys({ecc, Props}),
-            ECCWorker = [?WORKER(miner_ecc_worker, [KeySlot])];
-        {PublicKey, ECDHFun, SigFun} ->
-            ECCWorker = [],
-            ok
-    end,
-
-    BlockchainOpts = [
-        {key, {PublicKey, SigFun, ECDHFun}},
-        {seed_nodes, SeedNodes ++ SeedAddresses},
-        {max_inbound_connections, MaxInboundConnections},
-        {port, Port},
-        {num_consensus_members, NumConsensusMembers},
-        {base_dir, BaseDir},
-        {update_dir, application:get_env(miner, update_dir, undefined)},
-        {group_delete_predicate, fun miner_consensus_mgr:group_predicate/1}
-    ],
-
-    %% Miner Options
-    POCOpts = #{
-        base_dir => BaseDir
-    },
-
-    OnionServer =
-        case application:get_env(miner, radio_device, undefined) of
-            {RadioBindIP, RadioBindPort, RadioSendIP, RadioSendPort} ->
-                %% check if we are overriding/forcing the region ( for lora )
-                RegionOverRide = check_for_region_override(),
-                OnionOpts = #{
-                    radio_udp_bind_ip => RadioBindIP,
-                    radio_udp_bind_port => RadioBindPort,
-                    radio_udp_send_ip => RadioSendIP,
-                    radio_udp_send_port => RadioSendPort,
-                    ecdh_fun => ECDHFun,
-                    sig_fun => SigFun,
-                    region_override => RegionOverRide
+                 strategy => rest_for_one,
```

Comment:
so if a process under the critical sup goes down, the miner stops.  Seems strange to have that as the desired outcome.  why not allow the entire tree to restart rather than rely on external intervention ( canary or human ) ?

Reply:
that's the current behavior, and for whatever reason it doesn't seem to work particularly well.  this is definitely a bandaid.

Reply:
well not exactly...i think i commented on the wrong line originally so i maybe didnt convey things well...

The old top level sup `miner_sup` had a restart strategy of 

intensity => 2,
period => 5

The new has

intensity => 0,
period => 1

And as the miner_critical_sup has the same new restart strategy... if any of those processes under the critical sup, such as blockchain_worker, go down then both the critical sup and the miner sup restart strategies are breached taking down the miner.

In the old way for the same scenario the restart strategy would at least attempt to restart the miner tree 2 times.  Now there are no automatic restarts off the miner tree.  

Is that the desired new behaviour ?

Reply:
oh i have just realised the sups here are using a rest for one and were doing so in the original which meant the entire tree was unlikely to ever be restarted unless there was a breach of the restart policy.

if you changed the miner_sup to a one_for_all and a restart strategy of 5, 2 it would ensure the entire miner tree is restarted if something in the critical_sup goes down.

I dont know about the exact issue/scenario are you trying to address here but it could be the rest_for_one is resulting in some non clean behaviour during one of more process crashes..

Reply:
i tested the above approach locally and it works fine ( assuming you want the entire miner_sup tree restarted if something in the critical_sup goes down )...
Works fine except for one major caveat, during the restart the swarm sup fails to start and reports an already started error...so something is not right with how these processes are being supervised.

If you put the miner_sup intensity way up ( say 200 ), it eventually takes care of itself as the swarm_sup eventually dies and permits the restart of the miner tree

I am pretty sure i fixed this in one of my old unmerged prs addresses shutdown behaviour....i can have a look at it again later.

This lack of clean shutdown may be part of the root issue that was the driver behind this PR in the first place ?

Reply:
the swarm is started from blockchain_swarm and killed there also as part of blockchain_swarm's terminate
but as blockchain_swarm is not trapping exits the terminate is not called during a shutdown.
Adding a trap exit here gives a clean shutdown and enables the full miner_tree to be restarted cleanly in the event of one of the critical_sup's processes going down and avoids the already_started error

Reply:
i should prob revisit those clean shutdown PRs and get them up to date and we should try to get them merged

Reply:
To summarise my comments I am suggesting the following additional changes:

1. Make miner_sup a one_for_all type with a restart strategy of 5, 2
2. Update blockchain_swarm.erl to trap exits to ensure its terminate is called during a shutdown, restart etc which will ensure the swarm sup is stopped cleanly

Together this will ensure that should a process under the criticial_sup go down the entire miner tree will be restarted cleanly